### ISSUE TITLE
Possible Bug in the Recurrent Neural Network (RNN) implementation

### ISSUE THREAD

**Issue by zsogitbe** [mlpack/mlpack#2713]
#### Issue description

While training a RNN like this
```
for (int i = 0; i <= CYCLES; i++)
{
   model.Train(...
   model.Predict(...
}
```
the prediction step increases the size of the final model. If you use the training data for prediction then the size of the final model may become several times bigger just because of the Prediction step. For example, with a small test dataset in the prediction step you get a final model size of 2MB and if you use the full training dataset you get 20MB. It is not logical to me why the prediction step increase the size of the trained model.

#### Your environment

 * version of mlpack: last
 * operating system: Windows 10
 * compiler: VS2017
 * version of dependencies (Boost/Armadillo): last
 * any other environment information you think is relevant: -

#### Steps to reproduce

see above

#### Expected behavior

The Prediction step should not influence the size of the model.

#### Actual behavior

see above

**Comment by rcurtin** [comment]
@zsogitbe thanks for the report.  I think this may be very simple to fix.  @zoq should we clear `inputParameter` and `outputParameter` and other internal storage of the FFN at the end of `Predict()`?  (and RNN too I assume)

**Comment by zoq** [comment]
Yes, that sounds like a good approach.

**Comment by rcurtin** [comment]
@zsogitbe can you tell us about the network structure you are using?  I actually think my initial diagnosis was not correct here---I think some layers may be (incorrectly) caching data, but I want to make sure I actually catch the situation you're reporting.  If you can paste some code that, when run, reproduces the issue, that would be best. :+1:

**Comment by zsogitbe** [comment]
Hi Ryan, you can use any of the RNN examples (stock or electricity for this). You will need to use the 'old' way of training per cycle with which you can have a prediction step after each cycle. We need this prediction step in order to see the evolution of the MSQ (regression) or accuracy (classification) and not only the 'loss'... Case 1: train the model without a prediction step and save the model. Case 2: train the model with a prediction step (test) in each cycle and by using the whole training dataset. You will see that the model size in Case 1 will be very small and in Case 2 it will be several MB's larger. If you save the model in text form instead of binary then you will see that the prediction steps add the multiple of weights to the model. The prediction step must not modify the model in any way. I hope that this helps to track down the problem.

**Comment by zoq** [comment]
Thanks for the report, I'll take a look into the issue.
